Bluetooth: scan: Implement non-blocking ext-adv report reassembly

jori-nordic opened the pull request:
Re-assemble the extended advertising reports into a dedicated buffer
pool and notify the application asynchronously.

To achieve this we:
- decompose the re-assembly function into fragmented and non-frag parts
- for reassembled reports: push any other info onto the reassembled buffer
- for non-frag reports: pass the raw event buffer directly
- put that buffer on a FIFO
- trigger a (sys)work item that pulls and calls the cb

The optimization for 80% of cases is kept: Processing non-fragmented
extended advertising reports will not double-copy: We add a buffer pool
similar to `sync_evt_pool` and apply the following algo:
- if report is non-frag & we have at least 1 buf available in pool
  -> ref() that buffer and put it on the FIFO directly
- else:
  -> allocate from re-assembly pool and copy into it

Note that we can only apply this optimization when we are processing the
last report in the event buffer.

Note to reviewers: The diff of this commit is garbage, you're probably
better of comparing with the previous implementation manually.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>
Co-authored-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/50786

alwa-nordic committed:
Bluetooth: host: Alter `bt_hci_evt_get_flags` to take 3 bytes

This is done in preparation to select flags based on the LE meta event
code.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

jori-nordic committed:
Bluetooth: tests: host/scan/slow: run on nRF53

This enables the bluetooth/host/scan/slow test on nRF53 in the Zephyr
Continuous Integration pipeline.

Why waste more CI cycles?

The 5340 build will:
1. stress-test the HCI IPC driver (controller + host sides)
2. use a different buffer allocation model, where the allocator doesn't
know what other events are in the queue

2. is especially important, as this is the model a lot of host users
will be running with. Any controller driver that is not directly on-chip
_cannot_ know what is in the HCI queues, and cannot take the same
convenient shortcuts as the on-chip function-called drivers.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

alwa-nordic committed:
Bluetooth: host: Process adv report in prio

Since advertising report event processing defers the application
callback to a different context, it's non-blocking and should be
performed on in prio context to avoid blocking prio events.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

jori-nordic committed:
Bluetooth: buf: Allocate HCI events from sync_evt_pool

We allocate extended advertising reports from the `sync_evt_pool`. That
pool will always have a buffer available when `bt_recv()` is called.

This necessitates a new API for the drivers to get a buffer, as the old
one does not have enough information to detect extended advertising
reports.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jori-nordic committed:
Bluetooth: Make H(ost)Top

Sorry we don't have color support yet..

On the non-joke side, this option will print the number of dropped
extended advertising reports every second.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>
Co-authored-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

jori-nordic committed:
Bluetooth: tests: host/scan/slow: Validate advertising data

Add some advertising data and validate it on the peer.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jori-nordic commented:
There may be some funnies left in there. Do report and we will fix.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -2927,13 +2928,16 @@ static const struct event_handler normal_events[] = {
  * @return HCI event flags for the specified event.
  */
-static inline uint8_t bt_hci_evt_get_flags(uint8_t evt)
+static inline uint8_t bt_hci_evt_get_flags(uint8_t evt[static 3])
```
I've never seen the `static` keyword used in an array size designator like that. What kind of sorcery is this?

jhedberg replied:
And more importantly, which C standard is this from, since there's a risk that it may introduce some compatibility issues (see https://docs.zephyrproject.org/latest/develop/languages/c/index.html#language-standards)

jori-nordic replied:
@alwa-nordic is the C language lawyer lol.

another nugget was we had `while (num_reports --> 0)`. Which is also valid 🙃

jhedberg replied:
Looks like it's from C11, so we should be fine there: https://stackoverflow.com/questions/3430315/what-is-the-purpose-of-static-keyword-in-array-parameter-of-function-like-char

aescolar replied:
https://hamberg.no/erlend/posts/2013-02-18-static-array-indices.html :)

aescolar replied:
> @alwa-nordic is the C language lawyer lol.
> 
> another nugget was we had `while (num_reports --> 0)`. Which is also valid 🙃

perfectly valid code: `value += (value = value == value --> value <-- value), value++;` I see myself out now

jhedberg commented on include/zephyr/bluetooth/buf.h:
```diff
@@ -161,6 +161,8 @@ static inline enum bt_buf_type bt_buf_get_type(struct net_buf *buf)
 		->type;
 }
 
+struct net_buf *bt_buf_get_evt_but_better(uint8_t evt, uint8_t meta, k_timeout_t timeout);
```
I suppose we'd want to deprecate `bt_buf_get_evt()` as well? That said, you'd need to update all upstream HCI drivers first, I suppose.

jori-nordic replied:
yes. I made this to allow a soft-migration. Now that I think of it, I should update `bt_buf_can_steal_ext_adv_report_buf` to take `buf` and double-check the pool too. Because it won't work if the driver uses a different pool.

jori-nordic replied:
fixed. Found (and fixed) a bug in the process, yay.

jhedberg commented on subsys/bluetooth/host/buf.c:
```diff
@@ -82,6 +82,96 @@ struct net_buf *bt_buf_get_rx(enum bt_buf_type type, k_timeout_t timeout)
+	 * - There are buffers in the `free` LIFO
+	 * - There are un-initialized buffers
+	 */
+	return (!k_fifo_is_empty(&sync_adv_pool.free)) || (sync_adv_pool.uninit_count != 0);
```
This almost begs for a proper net_buf_pool API to test for "is it empty?". That said, it'd inherently racy in premptible contexts, so we might not want to give people such a foot gun.

jori-nordic replied:
yep. Ideally `net_buf` would init all buffers on boot and place them in the `free` LIFO.

jori-nordic commented:
@jhedberg @Thalley 
There is another issue, due to a missing half-bit in the spec:

Given that sequence:
```
[ Ai | Ai | Bi | Bc | Ac ]
```
Where `i` is incomplete-more-data and c is `complete`. `A` and `B` are two different peers.

What callbacks and data should we expect as a user? The current behavior (as tested by [`host_long_adv_recv`](https://github.com/jori-nordic/zephyr/blob/76d9699330a073ec11f66d7dc4bab43b3e01cb7f/tests/bluetooth/host_long_adv_recv/src/main.c#L487-L488)) is that we will receive the full data for `B`. BUT we also receive `A`, with the beginning truncated.

Do we ask for a new spec? <sup><sub>with a new status `START` using up the last half-bit</sub></sup>

hermabe commented:
> @jhedberg @Thalley There is another issue, due to a missing half-bit in the spec:
> 
> Given that sequence:
> 
> ```
> [ Ai | Ai | Bi | Bc | Ac ]
> ```
> 
> Where `i` is incomplete-more-data and c is `complete`. `A` and `B` are two different peers.
> 
> What callbacks and data should we expect as a user? The current behavior (as tested by [`host_long_adv_recv`](https://github.com/jori-nordic/zephyr/blob/76d9699330a073ec11f66d7dc4bab43b3e01cb7f/tests/bluetooth/host_long_adv_recv/src/main.c#L487-L488)) is that we will receive the full data for `B`. BUT we also receive `A`, with the beginning truncated.
> 
> Do we ask for a new spec? with a new status `START` using up the last half-bit

Throwback: https://github.com/zephyrproject-rtos/zephyr/pull/39352#issuecomment-942103833

Yes, the spec is broken. There is no way to know if you lost the beginning of the adv report. We have to rely on the controller not interleaving fragmented reports. We tried raising an errata last time and it didnt go anywhere.

hermabe commented:
https://bluetooth.atlassian.net/jira/software/c/projects/ES/issues/ES-17839

The suggested solution is to keep a list of advertiser address and data status. Completely ignoring that this data needs to be stored for an unbounded number of advertisers, and that one advertiser may have multiple sets and anonymous advertising exists.

Thalley commented on subsys/bluetooth/host/scan.c:
```diff
@@ -765,154 +867,266 @@ static void create_ext_adv_info(struct bt_hci_evt_le_ext_advertising_info const
+{
+	struct net_buf *b;
+
+	if (remaining_reports == 1 && bt_buf_can_steal_ext_adv_report_buf(*buf)) {
+		/* optimization, we can zero-copy instead of malloc */
+		b = net_buf_ref(*buf);
+
+		/* The caller should _not_ modify `buf` anymore. We stole it. */
+		*buf = NULL;
+	} else {
+		b = net_buf_alloc(&reassembly_pool, K_NO_WAIT);
+	}
```
I think I'm missing some detail. What the does the number of reports have to do with reassembly? Why can't we zero-copy when `remaining_reports = 2`?

jori-nordic replied:
If you were in oslo I could show you my sketches.
Basically, the optimization becomes more involved if we don't only apply it to the last report. We can only steal the `buf` from the HCI driver once. 
So if we want to steal `buf` for report number=1 out of num_reports=2, we have to:
- read from a `buf_simple` or something (and then fn signatures get really complicated)
- somehow ref `buf`, keep some state that we want to schedule it later
- read report 2, alloc and queue it
- remember we wanted to schedule report 1, fast-forward the `buf->data` pointer to report 1's location (which might not be the 1st report in real life)
- add the magic to report 1, and queue it

So it's quite a bit more code and control flow, and I'm unconvinced that we gain much from it. I think in 80% of cases we don't batch reports anyways and we'll have lots of `num_reports=1` with `COMPLETE`. This optimization is designed only for that case. The rest can "malloc", it's easier to follow.

As I said though, "I think". It would be good to collect some HCI traces at the UPF to see if SDC or other controllers tend to batch reports or send lots of small events. Then we can revisit.

Thalley replied:
Could I bother you or @alwa-nordic to put that explanation as a in-code comment? 

#thecodeisnotselfdocumenting

jori-nordic replied:
I'll definitely do that. It's always hard to put oneself in the reader's shoes

Thalley commented:
> https://bluetooth.atlassian.net/jira/software/c/projects/ES/issues/ES-17839
> 
> The suggested solution is to keep a list of advertiser address and data status. Completely ignoring that this data needs to be stored for an unbounded number of advertisers, and that one advertiser may have multiple sets and anonymous advertising exists.

Indeed, broken spec. Nothing much we can do about that. 

We can only do best effort and keep incomplete advertiser data for a specific amount of time before discarding it

jori-nordic commented on subsys/bluetooth/host/scan.c:
```diff
@@ -765,154 +867,266 @@ static void create_ext_adv_info(struct bt_hci_evt_le_ext_advertising_info const
-		data_status = BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS(evt_type);
-		is_report_complete = data_status == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_COMPLETE;
-		more_to_come = data_status == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_PARTIAL;
+static void crash_if_corrupted(size_t report_length, size_t buffer_length)
```
Yes we've had `__ASSERT()`. But what about second `BT_ASSERT()`?
BT_ASSERT is `default y` and decoupled from `CONFIG_ASSERT`. You have to go out of your way to disable it, the vast majority of users don't.

The function name doesn't live up to the impl, only if people deliberately shoot themselves in the foot, no.

jori-nordic commented on subsys/bluetooth/host/scan.c:
```diff
@@ -765,154 +867,266 @@ static void create_ext_adv_info(struct bt_hci_evt_le_ext_advertising_info const
+{
+	/* I hope it's a k_oops! */
+#define BT_HOPE BT_ASSERT_MSG
 
-		if (evt->length > buf->len) {
-			LOG_WRN("Adv report corrupted (wants %u out of %u)", evt->length, buf->len);
+	BT_HOPE(buffer_length >= report_length, "Adv report corrupted (wants %u out of %u)",
```
just for fun :) 
will remove

jori-nordic commented on subsys/bluetooth/host/scan.c:
```diff
@@ -765,154 +867,266 @@ static void create_ext_adv_info(struct bt_hci_evt_le_ext_advertising_info const
-			goto cont;
-		}
+	bool complete = data_status == COMPLETE;
+	bool borked = (data_status == INCOMPLETE_DATA_TRUNCATED_NO_MORE_TO_COME) ||
+		      (data_status == CONTACT_SALES);
```
I also have no idea. Its RFU :p
We should open a motion to the SIG to actually use this bit for `START_OF_CHAIN` or something actually useful.

Thalley replied:
Didn't we already open an erratum for this mess which was rejected? :(

jori-nordic replied:
we should open it again. @theob-pro  could you talk to Pål about it?

jori-nordic commented on subsys/bluetooth/host/scan.c:
```diff
@@ -765,154 +867,266 @@ static void create_ext_adv_info(struct bt_hci_evt_le_ext_advertising_info const
+	if (!discarding) {
+		__ASSERT_NO_MSG(r->report);
+		net_buf_add_mem(r->report, /* flex the member 🤖💪 */ evt->data, evt->length);
+	}
+
+	if (!discarding && complete) {
```
I think we do need the `!discarding`. Else we will enqueue `r->report` which is `NULL` when we are discarding. We can of course rework the logic to have this condition if we really want to.

Thalley replied:
My point here was that we already have a `!discarding` clause and the `complete` check could be moved into the existing `!discarding` check so we don't have to check that twice

jori-nordic replied:
I was reading your comment from vscode and it butchered the suggestion, sorry about that.
I tend to favor less indentation if possible but won't maintain this code. So it's up to @alwa-nordic and @jhedberg to decide.

jhedberg replied:
@jori-nordic @Thalley I'm also in general in favor of less indentation, either through playing around with conditions and function return locations, or by refactoring parts of the code to helper functions.

Thalley replied:
If we can assume that the compiler can optimize it so that

```c
if (foo) {
   bar();
}

if (foo && zoo) {
   zar();
}
```

Generates the same binary as 

```c
if (foo) {
    bare();
    
    if (zoo) {
        zar();
    }
}
```

Then I'm absolutely fine with it :) If `foo` needs to be evaluated twice for the sake of readability, then we are effectively saying that performance is less important than readability, which can be a slippery slope in some cases :)

jori-nordic commented on subsys/bluetooth/host/scan.c:
```diff
@@ -765,154 +867,266 @@ static void create_ext_adv_info(struct bt_hci_evt_le_ext_advertising_info const
+	}
+
+	if (is_part_of_buf(evt, report)) {
+		/* Oh my god karen, you can't just ask a buf to copy to itself */
```
Failed attempt at re-using https://knowyourmeme.com/memes/oh-my-god-karen-you-cant-just-ask-someone-why-theyre-white

Before, I had only the `else` branch. If `evt` was part of `buf` and I still did the memcpy, fun ensued.

jori-nordic replied:
Come to think of it, the assert could be placed below the if/else blocks

jori-nordic commented on subsys/bluetooth/host/scan.c:
```diff
@@ -765,154 +867,266 @@ static void create_ext_adv_info(struct bt_hci_evt_le_ext_advertising_info const
+		__ASSERT_NO_MSG((void *)(evt) == (void *)(report->data));
+	} else {
+		net_buf_reserve(report, REPORT_HEADER_SIZE);
+		/* slightly unsafe */
+		net_buf_add_mem(report, evt, sizeof(*evt) + evt->length);
```
I dislike using flexible members in general. Don't really think it's unsaf-er than other C. alwa had the idea to use it.

jori-nordic commented on subsys/bluetooth/host/scan.c:
```diff
@@ -765,154 +867,266 @@ static void create_ext_adv_info(struct bt_hci_evt_le_ext_advertising_info const
+
+void bt_hci_le_adv_ext_report(struct net_buf *reports)
+{
+	/* - Read-onlies `buf`
```
read-only is a verb, right? :p

jori-nordic replied:
This comment is outdated though. E.g. it doesn't return anything, `buf` is def not read-only anymore etc. I'll remove.

jori-nordic commented on subsys/bluetooth/host/scan.c:
```diff
@@ -56,40 +57,141 @@ static sys_slist_t scan_cbs = SYS_SLIST_STATIC_INIT(&scan_cbs);
-/* A buffer used to reassemble advertisement data from the controller. */
-NET_BUF_SIMPLE_DEFINE(ext_scan_buf, CONFIG_BT_EXT_SCAN_BUF_SIZE);
+
+enum __packed report_magic {
+	COMPLETE_REPORT_MAGIC = 0xbf,
+	REASSEMBLED_REPORT_MAGIC = 0xfb,
+};
+BUILD_ASSERT(sizeof(enum report_magic) == sizeof(uint8_t));
+
+#define REPORT_HEADER_SIZE                                                                         \
+	(sizeof(enum report_magic) + sizeof(bt_addr_le_t) + sizeof(struct bt_le_scan_recv_info))
```
It definitely does.

jori-nordic commented on subsys/bluetooth/host/scan.c:
```diff
@@ -56,40 +57,141 @@ static sys_slist_t scan_cbs = SYS_SLIST_STATIC_INIT(&scan_cbs);
+
+	k_work_submit(work);
+
+	if (do_not_want()) { /* it's an old meme but it checks out */
+		goto yeet;
+	}
+
+	uint8_t magic = net_buf_pull_u8(report);
+
+	if (magic == REASSEMBLED_REPORT_MAGIC) {
+		process_reassembled_report(report);
+	} else if (magic == COMPLETE_REPORT_MAGIC) {
+		process_unparsed_reassembled_report(report);
+	} else {
+		__ASSERT(0, "no fun without magic");
+		goto yeet;
+	}
+
+yeet:
+	net_buf_unref(report);
```
but then we lose the comedic effect :/

jori-nordic commented on subsys/bluetooth/host/scan.c:
```diff
@@ -56,40 +57,141 @@ static sys_slist_t scan_cbs = SYS_SLIST_STATIC_INIT(&scan_cbs);
-/* A buffer used to reassemble advertisement data from the controller. */
-NET_BUF_SIMPLE_DEFINE(ext_scan_buf, CONFIG_BT_EXT_SCAN_BUF_SIZE);
+
+enum __packed report_magic {
+	COMPLETE_REPORT_MAGIC = 0xbf,
+	REASSEMBLED_REPORT_MAGIC = 0xfb,
+};
```
huh, didn't even realize that.

jori-nordic commented on subsys/bluetooth/host/scan.c:
```diff
@@ -765,154 +867,266 @@ static void create_ext_adv_info(struct bt_hci_evt_le_ext_advertising_info const
-		bool is_report_complete;
-		bool more_to_come;
-		bool is_new_advertiser;
+		k_sleep(K_SECONDS(2));
```
I guess so. Next PR?

Thalley replied:
That would be OK for, but preferably in this one

jori-nordic commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -4054,6 +4070,18 @@ void hci_event_prio(struct net_buf *buf)
+	return &k_sys_work_q;
+#elif defined(CONFIG_BT_RECV_WORKQ_BT)
+	return &bt_workq;
+#else
+	BUILD_ASSERT("config not supported");
+	return NULL;
+#endif /* CONFIG_BT_RECV_WORKQ_SYS */
```
It was more a safeguard against downstreams meddling with kconfigs. I'll remove it, as the only two choices are sys and rx anyways.

jori-nordic commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -2947,6 +2951,13 @@ static inline uint8_t bt_hci_evt_get_flags(uint8_t evt)
 	case BT_HCI_EVT_CMD_STATUS:
 		return BT_HCI_EVT_FLAG_RECV_PRIO;
+	case BT_HCI_EVT_LE_META_EVENT:
+		switch (evt[2]) {
```
@alwa-nordic can do the needful. The drivers do this all the time however.

jori-nordic commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -1035,6 +1048,12 @@ config BT_TESTING
 	  This option enables custom Bluetooth testing interface.
 	  Shall only be used for testing purposes.
 
+config BT_HTOP
+	bool "Bluetooth Statistics"
+	help
+	  Enable printing of Bluetooth performance statistics.
+	  Sort of like THREAD_ANALYZER.
```
`htop` is better than `top`. I'll let you and @alwa-nordic + @jhedberg decide on a proper name.
Note that this commit could def be yeeted out to its own PR.

jori-nordic commented on subsys/bluetooth/host/buf.c:
```diff
@@ -82,6 +82,101 @@ struct net_buf *bt_buf_get_rx(enum bt_buf_type type, k_timeout_t timeout)
+	return (!k_fifo_is_empty(&sync_adv_pool.free)) || (sync_adv_pool.uninit_count != 0);
+}
+
+struct net_buf *bt_buf_get_evt_but_better(uint8_t evt, uint8_t meta, k_timeout_t timeout)
```
Not all users. Eg. `ipc.c` uses `SECONDS(60)`.
Some other `drivers/bluetooth/hci` do this too. Btw, we should probably update them too in a future PR.

jori-nordic committed:
Bluetooth: scan: Implement non-blocking ext-adv report reassembly

Re-assemble the extended advertising reports into a dedicated buffer
pool and notify the application asynchronously.

To achieve this we:
- decompose the re-assembly function into fragmented and non-frag parts
- for reassembled reports: push any other info onto the reassembled buffer
- for non-frag reports: pass the raw event buffer directly
- put that buffer on a FIFO
- trigger a (sys)work item that pulls and calls the cb

The optimization for 80% of cases is kept: Processing non-fragmented
extended advertising reports will not double-copy: We add a buffer pool
similar to `sync_evt_pool` and apply the following algo:
- if report is non-frag & we have at least 1 buf available in pool
  -> ref() that buffer and put it on the FIFO directly
- else:
  -> allocate from re-assembly pool and copy into it

Note that we can only apply this optimization when we are processing the
last report in the event buffer.

Note to reviewers: The diff of this commit is garbage, you're probably
better of comparing with the previous implementation manually.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>
Co-authored-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

Thalley commented on subsys/bluetooth/host/scan.c:
```diff
@@ -765,154 +872,285 @@ static void create_ext_adv_info(struct bt_hci_evt_le_ext_advertising_info const
+		return "INCOMPLETE-NM";
+	case STATUS_RFU:
+		return "RFU";
+	default:
+		__ASSERT(0, "Unknown data status %x", status);
+		return "";
+	}
+}
 
-		is_new_advertiser = reassembling_advertiser.state == FRAG_ADV_INACTIVE ||
-				    !fragmented_advertisers_equal(&reassembling_advertiser,
-								  &evt->addr, evt->sid);
+static bool process_fragmented_report(struct bt_hci_evt_le_ext_advertising_info *evt)
+{
+	/* This function parses the event and re-assembles fragmented reports
+	 * into synthetic events.
+	 *
+	 * Returns true if the event was processed. I.e. will return false for
+	 * reports that fit a single event and are complete.
+	 */
+	struct fragmented_advertiser *r = &reassembling_advertiser;
 
-		if (is_new_advertiser && is_report_complete) {
-			/* Only advertising report from this advertiser.
-			 * Create event immediately.
-			 */
-			create_ext_adv_info(evt, &scan_info);
-			le_adv_recv(&evt->addr, &scan_info, &buf->b, evt->length);
-			goto cont;
-		}
+	uint16_t evt_type = sys_le16_to_cpu(evt->evt_type);
+	uint16_t data_status = BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS(evt_type);
```
`status_to_str` is called with `data_status` which has the return value of `BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS`. 

However the values returned by `BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS` are not the same values as `status_to_str` matches on. 

`BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS` should return one of the following: 

```c
BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_COMPLETE
BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_PARTIAL
BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_INCOMPLETE
```

Thalley commented on subsys/bluetooth/host/scan.c:
```diff
@@ -765,154 +872,285 @@ static void create_ext_adv_info(struct bt_hci_evt_le_ext_advertising_info const
+	}
+
+	if (!discarding && complete) {
+		/* Core Spec v5.4. 4.E.7.7.65.13 says controller uses latest
+		 * RSSI. Host wants to be like controller when grow up.
```
> Host wants to be like controller when grow up.

Can you elaborate on what that means?

narvalotech replied:
Hey, I'll comment with my personal account :)

@alwa-nordic should rephrase it. Basically, it means that the host should not have a surprising behavior: if the spec has a guideline for the controller, the host should also (unless otherwise stated) try to follow it.

Here the guideline for the LL is that it should use the RSSI from the latest AUX packet received for enqueuing the HCI ext-adv-report event. In the host we translate that to "use the RSSI value from the last ext-adv-report fragment in the chain when you notify the application".
I hope that makes sense.

Thalley replied:
Ah, that makes sense.

jhedberg commented:
@alwa-nordic do you intend to pick up the work here, or is @jori-nordic / @narvalotech still going to work on it?

alwa-nordic commented:
I will pick up the work.
